shopware/frontends
# poc: session context from nuxt plugin

## How it works
PoC Exposes a current session in computed property in a nuxt plugin level (client&server) `const { $getCurrentSession } = useNuxtApp()`

it's not a final state, only a proposition that may solve an issue like described [here](https://github.com/shopware/frontends/issues/641).

This little change move logic responsible for fetching a context to the lower level of a nuxt app (from app.vue to nuxt plugin). thanks to this, session context can be read in a nuxt middleware, for instance. 

## Why?
The current session is the data frequently needed from the very beginning state of a running application, when nuxt context is not set, vue lifecycle hooks are not available.

- allows us to use the current session according to the `sw-context-token` provided in the cookie (remains stateless - only cookie is used, no other storage is needed)
- it saves requests, in comparison to using the plain `apiClient` instance in the middleware
- does not require changes in the other composables API
- does not make a singleton of any reference - bases only on a plugin

## What does it contain?

- moved logic of fetching a context from `app.vue` to nuxt3-module's plugin
- in SSR it fetches the context only once, then pass it via ssrContext.payload (similar way of how it's managed when using `useAsyncData`)
- the session is being read on the client side (also in a plugin) if it had been set in SSR, otherwise it's being loaded using api client
- fixes unnecessary language patching

## TEST IT:
- run app and go to `/shopware` route to see the middleware working https://frontends-demo-git-poc-session-contex-26fff6-shopware-frontends.vercel.app/shopware
- log in 
- reload the page
- log out
- etc...

## Next steps
- Could be also a good solution for the another layer of composables - nuxt composables when `useNuxtApp` is available - then use this approach, when it's a pure vue, use composition api for vue context only.
- check if works in CSR mode only
- check possibly memory leaks
